File: cminus/Machine.fs
(* File MicroC/Machine.fs 

   Instructions and code emission for a stack-based
   abstract machine * sestoft@itu.dk 2009-09-23

   Implementations of the machine are found in file MicroC/Machine.java 
   and MicroC/machine.c.
   
   Must precede Comp.fs and Contcomp.fs in the VS Solution Explorer.
 *)


module Machine

type label = string
// 汇编指令
  
type instr =
  | Label of label                     (* symbolic label; pseudo-instruc. *)
  | CSTI of int                        (* constant                        *)
  | CSTF of int                        (* constant                        *)
  | CSTS of int list                        
  | ADD                                (* addition                        *)
  | SUB                                (* subtraction                     *)
  | MUL                                (* multiplication                  *)
  | DIV                                (* division                        *)
  | MOD                                (* modulus                         *)
  | EQ                                 (* equality: s[sp-1] == s[sp]      *)
  | LT                                 (* less than: s[sp-1] < s[sp]      *)
  | NOT                                (* logical negation:  s[sp] != 0   *)
  | DUP                                (* duplicate stack top             *)
  | SWAP                               (* swap s[sp-1] and s[sp]          *)
  | LDI                                (* get s[s[sp]]                    *)
  | STI                                (* set s[s[sp-1]]                  *)
  | GETBP                              (* get bp                          *)
  | GETSP                              (* get sp                          *)
  | INCSP of int                       (* increase stack top by m         *)
  | GOTO of label                      (* go to label                     *)
  | IFZERO of label                    (* go to label if s[sp] == 0       *)
  | IFNZRO of label                    (* go to label if s[sp] != 0       *)
  | CALL of int * label                (* move m args up 1, push pc, jump *)
  | TCALL of int * int * label         (* move m args down n, jump        *)
  | RET of int                         (* pop m and return to s[sp]       *)
  | PRINTI                             (* print s[sp] as integer          *)
  | PRINTC                             (* print s[sp] as character        *)
  | PRINTF                             (* print s[sp] as float            *)
  | LDARGS                             (* load command line args on stack *)
  | STOP                               (* halt the abstract machine       *)
  | BITAND
  | BITOR
  | BITXOR
  | BITLEFT
  | BITRIGHT
  | BITNOT
  | NEG
  | INVO
  | GCD
  | FLOOR
  | CEIL
  | ROUND
  | FTOI
  | ITOF
  | COS
  | TAN
  | ASIN
  | ACOS
  | ATAN
  | SIN
  | FABS
  | LOG
  | SQRT
  | POW
(* Generate new distinct labels *)

// 返回两个函数 resetLabels , newLabel
let (resetLabels, newLabel) = 
    let lastlab = ref -1
    ((fun () -> lastlab := 0), (fun () -> (lastlab := 1 + !lastlab; "L" + (!lastlab).ToString())))

(* Simple environment operations *)

type 'data env = (string * 'data) list

let rec lookup env x = 
    match env with 
    | []         -> failwith (x + " not found")
    | (y, v)::yr -> if x=y then v else lookup yr x

(* An instruction list is emitted in two phases:
   * pass 1 builds an environment labenv mapping labels to addresses 
   * pass 2 emits the code to file, using the environment labenv to 
     resolve labels
 *)

(* These numeric instruction codes must agree with Machine.java: *)



//机器码

//[<Literal>] 属性可以让 
//该变量在模式匹配时候被匹配,否则匹配时只能用数值.不能用变量名



[<Literal>]
let CODECSTI   = 0 


[<Literal>]
let CODEADD    = 1 


[<Literal>]
let CODESUB    = 2 


[<Literal>]
let CODEMUL    = 3 

[<Literal>]
let CODEDIV    = 4 

[<Literal>]
let CODEMOD    = 5 

[<Literal>]
let CODEEQ     = 6 

[<Literal>]
let CODELT     = 7 

[<Literal>]
let CODENOT    = 8 

[<Literal>]
let CODEDUP    = 9 

[<Literal>]
let CODESWAP   = 10 

[<Literal>]
let CODELDI    = 11 


[<Literal>]
let CODESTI    = 12 


[<Literal>]
let CODEGETBP  = 13 


[<Literal>]
let CODEGETSP  = 14 


[<Literal>]
let CODEINCSP  = 15 


[<Literal>]
let CODEGOTO   = 16


[<Literal>]
let CODEIFZERO = 17

[<Literal>]
let CODEIFNZRO = 18 


[<Literal>]
let CODECALL   = 19


[<Literal>]
let CODETCALL  = 20

[<Literal>]
let CODERET    = 21


[<Literal>]
let CODEPRINTI = 22 

[<Literal>]
let CODEPRINTC = 23


[<Literal>]
let CODELDARGS = 24

[<Literal>]
let CODESTOP   = 25;

[<Literal>]
let CODEBITAND   = 26;

[<Literal>]
let CODEBITOR   = 27;

[<Literal>]
let CODEBITXOR   = 28;

[<Literal>]
let CODEBITLEFT   = 29;

[<Literal>]
let CODEBITRIGHT  = 30;

[<Literal>]
let CODEBITNOT  = 31;

[<Literal>]
let CODENEG  = 32;

[<Literal>]
let CODEINVO  = 33;

[<Literal>]
let CODEGCD  = 34;
[<Literal>]
let CODEROUND  = 35;
[<Literal>]
let CODEFLOOR = 36;
[<Literal>]
let CODECEIL  = 37;
[<Literal>]
let CODECSTF  = 38;

[<Literal>]
let CODEPRINTF  = 39;

[<Literal>]
let CODEFTOI  = 40;

[<Literal>]
let CODEITOF  = 41;


[<Literal>]
let CODECOS  = 42;


[<Literal>]
let CODETAN  = 43;


[<Literal>]
let CODEASIN  = 44;


[<Literal>]
let CODEACOS  = 45;


[<Literal>]
let CODEATAN  = 46;


[<Literal>]
let CODESIN  = 47;

[<Literal>]

let CODEFABS  = 48;

[<Literal>]
let CODELOG  = 49;

[<Literal>]
let CODESQRT  = 50;

[<Literal>]
let CODEPOW  = 51;

[<Literal>]
let CODECSTS  = 52;

(* Bytecode emission, first pass: build environment that maps 
   each label to an integer address in the bytecode.
 *)
//获得标签在机器码中的地址
let makelabenv (addr, labenv) instr = 
    match instr with
    // 记录当前 (标签, 地址) ==> 到 labenv中
    | Label lab      -> (addr, (lab, addr) :: labenv)
    | CSTI i         -> (addr+2, labenv)
    | CSTF i         -> (addr+2, labenv)
    | CSTS i         -> (addr+2, labenv)
    | ADD            -> (addr+1, labenv)
    | SUB            -> (addr+1, labenv)
    | MUL            -> (addr+1, labenv)
    | DIV            -> (addr+1, labenv)
    | MOD            -> (addr+1, labenv)
    | EQ             -> (addr+1, labenv)
    | LT             -> (addr+1, labenv)
    | NOT            -> (addr+1, labenv)
    | DUP            -> (addr+1, labenv)
    | SWAP           -> (addr+1, labenv)
    | LDI            -> (addr+1, labenv)
    | STI            -> (addr+1, labenv)
    | GETBP          -> (addr+1, labenv)
    | GETSP          -> (addr+1, labenv)
    | INCSP m        -> (addr+2, labenv)
    | GOTO lab       -> (addr+2, labenv)
    | IFZERO lab     -> (addr+2, labenv)
    | IFNZRO lab     -> (addr+2, labenv)
    | CALL(m,lab)    -> (addr+3, labenv)
    | TCALL(m,n,lab) -> (addr+4, labenv)
    | RET m          -> (addr+2, labenv)
    | PRINTI         -> (addr+1, labenv)
    | PRINTC         -> (addr+1, labenv)
    | PRINTF         -> (addr+1, labenv)
    | LDARGS         -> (addr+1, labenv)
    | STOP           -> (addr+1, labenv)
    | BITAND         -> (addr+1, labenv)
    | BITOR          -> (addr+1, labenv)
    | BITXOR         -> (addr+1, labenv)
    | BITLEFT        -> (addr+1, labenv)
    | BITRIGHT       -> (addr+1, labenv)
    | BITNOT         -> (addr+1, labenv)
    | NEG            -> (addr+1, labenv)
    | INVO           -> (addr+1, labenv)
    | GCD            -> (addr+1, labenv)
    | FLOOR          -> (addr+1, labenv)
    | CEIL           -> (addr+1, labenv)
    | ROUND          -> (addr+1, labenv)
    | FTOI           -> (addr+1, labenv)
    | ITOF           -> (addr+1, labenv)
    | COS            -> (addr+1, labenv)
    | TAN            -> (addr+1, labenv)
    | ASIN           -> (addr+1, labenv)
    | ACOS           -> (addr+1, labenv)
    | ATAN           -> (addr+1, labenv)
    | SIN            -> (addr+1, labenv)
    | FABS           -> (addr+1, labenv)
    | LOG            -> (addr+1, labenv)
    | SQRT           -> (addr+1, labenv)
    | POW            -> (addr+1, labenv)
(* Bytecode emission, second pass: output bytecode as integers *)

//getlab 是得到标签所在地址的函数
//let getlab lab = lookup labenv lab

let rec emitints getlab instr ints = 
    match instr with
    | Label lab      -> ints
    | CSTI i         -> CODECSTI   :: i :: ints
    | CSTF i         -> CODECSTF   :: i :: ints
    | CSTS i         -> CODECSTS   :: i.[0] ::i.[1]::ints
    | ADD            -> CODEADD    :: ints
    | SUB            -> CODESUB    :: ints
    | MUL            -> CODEMUL    :: ints
    | DIV            -> CODEDIV    :: ints
    | MOD            -> CODEMOD    :: ints
    | EQ             -> CODEEQ     :: ints
    | LT             -> CODELT     :: ints
    | NOT            -> CODENOT    :: ints
    | DUP            -> CODEDUP    :: ints
    | SWAP           -> CODESWAP   :: ints
    | LDI            -> CODELDI    :: ints
    | STI            -> CODESTI    :: ints
    | GETBP          -> CODEGETBP  :: ints
    | GETSP          -> CODEGETSP  :: ints
    | INCSP m        -> CODEINCSP  :: m :: ints
    | GOTO lab       -> CODEGOTO   :: getlab lab :: ints
    | IFZERO lab     -> CODEIFZERO :: getlab lab :: ints
    | IFNZRO lab     -> CODEIFNZRO :: getlab lab :: ints
    | CALL(m,lab)    -> CODECALL   :: m :: getlab lab :: ints
    | TCALL(m,n,lab) -> CODETCALL  :: m :: n :: getlab lab :: ints
    | RET m          -> CODERET    :: m :: ints
    | PRINTI         -> CODEPRINTI :: ints
    | PRINTC         -> CODEPRINTC :: ints
    | PRINTF         -> CODEPRINTF :: ints
    | LDARGS         -> CODELDARGS :: ints
    | STOP           -> CODESTOP   :: ints
    | BITAND         -> CODEBITAND :: ints
    | BITOR          -> CODEBITOR  :: ints
    | BITXOR         -> CODEBITXOR :: ints
    | BITLEFT        -> CODEBITLEFT :: ints
    | BITRIGHT       -> CODEBITRIGHT :: ints
    | BITNOT         -> CODEBITNOT :: ints    
    | NEG            -> CODENEG  :: ints 
    | INVO           -> CODEINVO :: ints   
    | GCD            -> CODEGCD  :: ints
    | FLOOR          -> CODEFLOOR  :: ints
    | ROUND          -> CODEROUND  :: ints
    | CEIL           -> CODECEIL  :: ints
    | FTOI           -> CODEFTOI  :: ints
    | ITOF           -> CODEITOF  :: ints
    | COS            -> CODECOS   :: ints
    | TAN            -> CODETAN  :: ints
    | ASIN           -> CODEASIN  :: ints
    | ACOS           -> CODEACOS :: ints
    | ATAN           -> CODEATAN  :: ints
    | SIN            -> CODESIN  :: ints
    | FABS           -> CODEFABS  :: ints
    | LOG            -> CODELOG  :: ints
    | SQRT           -> CODESQRT  :: ints
    | POW            -> CODEPOW  :: ints

(* Convert instruction list to int list in two passes:
   Pass 1: build label environment
   Pass 2: output instructions using label environment
 *)
 
//通过对 code 的两次遍历,完成汇编指令到机器指令的转换
let code2ints (code : instr list) : int list =
    
    //从前往后遍历 `汇编指令序列 code: instr list`
    //得到 标签对应的地址,记录到 labenv中
    let (_, labenv) = List.fold makelabenv (0, []) code
    
    //getlab 是得到标签所在地址的函数
    let getlab lab = lookup labenv lab
    
    //从后往前 遍历 `汇编指令序列 code: instr list`
    List.foldBack (emitints getlab) code []
                    


let ntolabel (n:int) :label = 
    string(n)

//反编译
let rec decomp ints : instr list = 
    // printf "%A" ints
    match ints with
    | []                                              ->  []
    | CODEFABS :: ints_rest                           ->   FABS           :: decomp ints_rest
    | CODELOG :: ints_rest                            ->    LOG           :: decomp ints_rest
    | CODESQRT :: ints_rest                           ->   SQRT           :: decomp ints_rest
    | CODEPOW :: ints_rest                            ->    POW            :: decomp ints_rest
    | CODEITOF :: ints_rest                           ->   ITOF           :: decomp ints_rest
    | CODEFTOI :: ints_rest                           ->   FTOI           :: decomp ints_rest
    | CODEGCD  :: ints_rest                           ->   GCD           :: decomp ints_rest
    | CODEFLOOR  :: ints_rest                         ->   FLOOR            :: decomp ints_rest
    | CODECEIL  :: ints_rest                          ->   CEIL           :: decomp ints_rest
    | CODEROUND :: ints_rest                          ->   ROUND           :: decomp ints_rest
    | CODECOS:: ints_rest                             ->   COS           :: decomp ints_rest
    | CODETAN :: ints_rest                            ->   TAN           :: decomp ints_rest
    | CODEASIN  :: ints_rest                          ->   ASIN           :: decomp ints_rest
    | CODEACOS  :: ints_rest                          ->   ACOS            :: decomp ints_rest
    | CODEATAN  :: ints_rest                          ->   ATAN           :: decomp ints_rest
    | CODESIN   :: ints_rest                          ->   SIN           :: decomp ints_rest
    | CODEINVO :: ints_rest                           ->   INVO          :: decomp ints_rest
    | CODENEG :: ints_rest                            ->   NEG           :: decomp ints_rest
    | CODEBITLEFT :: ints_rest                        ->   BITLEFT       :: decomp ints_rest
    | CODEBITNOT :: ints_rest                         ->   BITNOT       :: decomp ints_rest
    | CODEBITRIGHT :: ints_rest                       ->   BITRIGHT      :: decomp ints_rest  
    | CODEBITAND :: ints_rest                         ->   BITAND        :: decomp ints_rest
    | CODEBITOR :: ints_rest                          ->   BITOR         :: decomp ints_rest
    | CODEBITXOR :: ints_rest                         ->   BITXOR        :: decomp ints_rest
    | CODEADD :: ints_rest                            ->   ADD           :: decomp ints_rest
    | CODESUB    :: ints_rest                         ->   SUB           :: decomp ints_rest
    | CODEMUL    :: ints_rest                         ->   MUL           :: decomp ints_rest
    | CODEDIV    :: ints_rest                         ->   DIV           :: decomp ints_rest
    | CODEMOD    :: ints_rest                         ->   MOD           :: decomp ints_rest
    | CODEEQ     :: ints_rest                         ->   EQ            :: decomp ints_rest
    | CODELT     :: ints_rest                         ->   LT            :: decomp ints_rest
    | CODENOT    :: ints_rest                         ->   NOT           :: decomp ints_rest
    | CODEDUP    :: ints_rest                         ->   DUP           :: decomp ints_rest
    | CODESWAP   :: ints_rest                         ->   SWAP          :: decomp ints_rest
    | CODELDI    :: ints_rest                         ->   LDI           :: decomp ints_rest
    | CODESTI    :: ints_rest                         ->   STI           :: decomp ints_rest
    | CODEGETBP  :: ints_rest                         ->   GETBP         :: decomp ints_rest
    | CODEGETSP  :: ints_rest                         ->   GETSP         :: decomp ints_rest
    | CODEINCSP  :: m :: ints_rest                    ->   INCSP m       :: decomp ints_rest
    | CODEGOTO   ::  lab :: ints_rest           ->   GOTO (ntolabel lab)      :: decomp ints_rest
    | CODEIFZERO ::  lab :: ints_rest           ->   IFZERO (ntolabel lab)    :: decomp ints_rest
    | CODEIFNZRO ::  lab :: ints_rest           ->   IFNZRO (ntolabel lab)    :: decomp ints_rest
    | CODECALL   :: m ::  lab :: ints_rest      ->   CALL(m, ntolabel lab)   :: decomp ints_rest
    | CODETCALL  :: m :: n ::  lab :: ints_rest ->   TCALL(m,n,ntolabel lab):: decomp ints_rest
    | CODERET    :: m :: ints_rest                    ->   RET m         :: decomp ints_rest
    | CODEPRINTI :: ints_rest                         ->   PRINTI        :: decomp ints_rest
    | CODEPRINTC :: ints_rest                         ->   PRINTC        :: decomp ints_rest
    | CODEPRINTF :: ints_rest                         ->   PRINTF       :: decomp ints_rest
    | CODELDARGS :: ints_rest                         ->   LDARGS        :: decomp ints_rest
    | CODESTOP   :: ints_rest                         ->   STOP             :: decomp ints_rest
    | CODECSTI   :: i :: ints_rest                    ->   CSTI i :: decomp ints_rest   
    | CODECSTF   :: i :: ints_rest                    ->   CSTF i :: decomp ints_rest 
    | CODECSTS   :: i :: j::ints_rest                 ->   CSTI i :: CSTI j::decomp ints_rest 
    | _                                       ->    printf "%A" ints; failwith "unknow code"

File: cminus/Machine.java
/* File MicroC/Machine.java
   A unified-stack abstract machine for imperative programs.
   sestoft@itu.dk * 2001-03-21, 2009-09-24

   To execute a program file using this abstract machine, do:

      java Machine <programfile> <arg1> <arg2> ...

   or, to get a trace of the program execution:

      java Machinetrace <programfile> <arg1> <arg2> ...

*/

import java.io.*;
import java.util.*;

class Machine {
  public static void main(String[] args)        
    throws FileNotFoundException, IOException {
    if (args.length == 0) 
      System.out.println("Usage: java Machine <programfile> <arg1> ...\n");
    else
      execute(args, false);
  }

  // These numeric instruction codes must agree with Machine.fs:

  final static int 
    CSTI = 0, ADD = 1, SUB = 2, MUL = 3, DIV = 4, MOD = 5, 
    EQ = 6, LT = 7, NOT = 8, 
    DUP = 9, SWAP = 10, 
    LDI = 11, STI = 12, 
    GETBP = 13, GETSP = 14, INCSP = 15, 
    GOTO = 16, IFZERO = 17, IFNZRO = 18, CALL = 19, TCALL = 20, RET = 21, 
    PRINTI = 22, PRINTC = 23, 
    LDARGS = 24,STOP = 25, BITAND = 26,BITOR = 27,BITXOR = 28,BITLEFT = 29,BITRIGHT = 30,
    BITNOT = 31,NEG = 32,INVO = 33,GCD = 34,ROUND=35,FLOOR=36,CEIL=37,CSTF=38,
    PRINTF = 39,FTOI=40 ,ITOF=41 ,COS  = 42,TAN  = 43,ASIN  = 44 ,ACOS  = 45,ATAN  = 46,SIN  = 47, FABS = 48,
    LOG = 49,SQRT = 50,POW = 51,CSTS=52;


  final static int STACKSIZE = 1000;
  
  // Read code from file and execute it

  static void execute(String[] args, boolean trace) 
    throws FileNotFoundException, IOException {
    int[] p = readfile(args[0]);                // Read the program from file
    int[] s = new int[STACKSIZE];               // The evaluation stack
    int[] iargs = new int[args.length-1];
    for (int i=1; i<args.length; i++)           // Push commandline arguments
      iargs[i-1] = Integer.parseInt(args[i]);
    long starttime = System.currentTimeMillis();
    execcode(p, s, iargs, trace);            // Execute program proper
    long runtime = System.currentTimeMillis() - starttime;
    System.err.println("\nRan " + runtime/1000.0 + " seconds");
  }

//a**b
  static int f(int a, int b){
    int result = 1;
    for(int i=1;i<=b;i++)
      result*=a;
    return result;
  }
  
  static int gcd(int a, int b){
    if(b==0)
      return a;
    return gcd(b,a%b);
  }
  // The machine: execute the code starting at p[pc] 
  static double change(int x){
      int tmp=x;
      x=Math.abs(x);
      int a=x%10000/100,b=x/10000;
      if(tmp>=0) 
        return (double) (b*Math.pow(10,-a));
      else return -(double) (b*Math.pow(10,-a));
  }
  static boolean judge(String x){
    if(x.endsWith("78")) 
      return true;
    return false;
  }
  static int change2(double x){
      String tmp=String.valueOf(x);
	    int index=tmp.indexOf("."),len;
	    // System.out.println(x);
	    if(index+3>tmp.length())len=tmp.length();else len=index+3;
      boolean flag=(x>=0);
      String tmp2=tmp.substring(0,len);
      double x2=Math.abs(Double.parseDouble(tmp2));
      tmp=String.valueOf(x2);
      index=tmp.indexOf(".");
	    // System.out.println(x2);
	    if(flag) 
	    	return ((int)(x2*Math.pow(10,tmp.length()-index-1))*100+tmp.length()-index-1)*100+78;
	    else
	    	return -(((int)(x2*Math.pow(10,tmp.length()-index-1))*100+tmp.length()-index-1)*100+78); 
	}
  static int execcode(int[] p, int[] s, int[] iargs, boolean trace) {
    int bp = -999;	// Base pointer, for local variable access 
    int sp = -1;	// Stack top pointer
    int pc = 0;		// Program counter: next instruction
    for (;;) {
      if (trace) 
        printsppc(s, bp, sp, p, pc);
      switch (p[pc++]) {   
      case CSTI:
        s[sp+1] = p[pc++]; sp++; break;
      case CSTF:
        s[sp+1] = p[pc++]; sp++; break;
      case CSTS:
        // System.out.println(p[pc-1]+" "+p[pc]+" "+p[pc+1]+" ");
        s[sp+1] = p[pc];  sp+=1;
        s[sp+1] = p[pc+1]; sp+=1;
        pc+=3;
        // System.out.print(p[pc]);
        break;

      case ADD:{
        boolean flag=false;double f1=s[sp-1],f2=s[sp];
        if(judge(String.valueOf(s[sp-1]))){f1=change(s[sp-1]);flag=true;}
        if(judge(String.valueOf(s[sp]))){f2=change(s[sp]);flag=true;}
        if(flag){ s[sp-1] =change2(f1+f2); sp--; break;}
        else{ s[sp-1] = s[sp-1] + s[sp]; sp--; break;}
      }
      case BITNOT: 
          s[sp] = ~s[sp]         ; break;
      case FABS:{
        double f=change(s[sp]);
        s[sp]=change2(Math.abs(f));
        break;
      }
      case LOG:{
        double f=change(s[sp]);
        s[sp]=change2(Math.log(f));
        break;
      }
      case SQRT:{
        double f=change(s[sp]);
        s[sp]=change2(Math.sqrt(f));
        break;
      }
      case POW:{
        double f1=change(s[sp-1]);
        double f2=change(s[sp]);
        s[sp-1]=change2(Math.pow(f1,f2));
        sp--;
        break;
      }
      case COS:{
        double f=change(s[sp]);
        s[sp]=change2(Math.cos(f));
        break;
      }
      case TAN:{
        double f=change(s[sp]);
        s[sp]=change2(Math.tan(f));
        break;
      }   
      case ASIN:{
        double f=change(s[sp]);
        s[sp]=change2(Math.asin(f));
        break;
      }
      case ACOS:{
        double f=change(s[sp]);
        s[sp]=change2(Math.acos(f));
        break;
      }   
      case ATAN:{
        double f=change(s[sp]);
        s[sp]=change2(Math.atan(f));
        break;
      }
      case SIN:{
        double f=change(s[sp]);
        s[sp]=change2(Math.sin(f));
        break;
      }   
      case ROUND:{
        double f=change(s[sp]);
        s[sp]=change2(Math.round(f));
        break;
      }
      case FLOOR:{
        double f=change(s[sp]);
        s[sp]=change2(Math.floor(f));
        break;
      }   
      case CEIL:{
        double f=change(s[sp]);
        s[sp]=change2(Math.ceil(f));
        break;
      }
      case FTOI:{
        double f=change(s[sp]);
        s[sp]=(int)f;
        break;
      }
      case ITOF:{
        s[sp]=change2(s[sp]);
        break;
      }
      case BITLEFT: 
          s[sp-1] = s[sp-1] << s[sp]; sp--; break;
      case BITRIGHT: 
          s[sp-1] = s[sp-1] >> s[sp]; sp--; break;
      case BITAND: 
          s[sp-1] = s[sp-1] & s[sp]; sp--; break;
      case BITOR: 
          s[sp-1] = s[sp-1] | s[sp]; sp--; break;
      case BITXOR: 
          s[sp-1] = s[sp-1] ^ s[sp]; sp--; break;
      case NEG:
          s[sp] = -s[sp];break;
      case INVO:
          s[sp-1] = f(s[sp-1], s[sp]); sp--; break;
      case GCD:
          int r;
          if(s[sp-1]<s[sp])
            r = gcd(s[sp],s[sp-1]);
          else 
            r = gcd(s[sp-1], s[sp]);
          s[sp-1] = r; sp--; break;
      case SUB:{
        boolean flag=false;double f1=s[sp-1],f2=s[sp];
        if(judge(String.valueOf(s[sp-1]))){f1=change(s[sp-1]);flag=true;}
        if(judge(String.valueOf(s[sp]))){f2=change(s[sp]);flag=true;}
        if(flag){ s[sp-1] =change2(f1-f2); sp--; break;}
        else{ s[sp-1] = s[sp-1] - s[sp]; sp--; break;}
      } 
      case MUL: {
        boolean flag=false;double f1=s[sp-1],f2=s[sp];
        if(judge(String.valueOf(s[sp-1]))){f1=change(s[sp-1]);flag=true;}
        if(judge(String.valueOf(s[sp]))){f2=change(s[sp]);flag=true;}
        if(flag){ s[sp-1] =change2(f1*f2); sp--; break;}
        else{  s[sp-1] = s[sp-1] * s[sp]; sp--; break;}
      }
      case DIV: {
        boolean flag=false;double f1=s[sp-1],f2=s[sp];
        if(judge(String.valueOf(s[sp-1]))){f1=change(s[sp-1]);flag=true;}
        if(judge(String.valueOf(s[sp]))){f2=change(s[sp]);flag=true;}
        if(flag){ s[sp-1] =change2(f1/f2); sp--; break;}
        else{  s[sp-1] = s[sp-1] / s[sp]; sp--; break;}
      }
      case MOD: 
        s[sp-1] = s[sp-1] % s[sp]; sp--; break;
      case EQ: 
        s[sp-1] = (s[sp-1] == s[sp] ? 1 : 0); sp--; break;
      case LT: 
        s[sp-1] = (s[sp-1] < s[sp] ? 1 : 0); sp--; break;
      case NOT: 
        s[sp] = (s[sp] == 0 ? 1 : 0); break;
      case DUP: 
        s[sp+1] = s[sp]; sp++; break;
      case SWAP: 
        { int tmp = s[sp];  s[sp] = s[sp-1];  s[sp-1] = tmp; } break; 
      case LDI:                 // load indirect
        s[sp] = s[s[sp]]; break;
      case STI:                 // store indirect, keep value on top
        s[s[sp-1]] = s[sp]; s[sp-1] = s[sp]; sp--; break;
      case GETBP:
        s[sp+1] = bp; sp++; break;
      case GETSP:
        s[sp+1] = sp; sp++; break;
      case INCSP:
        sp = sp+p[pc++]; break;
      case GOTO:
        pc = p[pc]; break;
      case IFZERO:
        pc = (s[sp--] == 0 ? p[pc] : pc+1); break;
      case IFNZRO:
        pc = (s[sp--] != 0 ? p[pc] : pc+1); break;
      case CALL: { 
        int argc = p[pc++];
        for (int i=0; i<argc; i++)	   // Make room for return address
          s[sp-i+2] = s[sp-i];		   // and old base pointer
        s[sp-argc+1] = pc+1; sp++; 
        s[sp-argc+1] = bp;   sp++; 
        bp = sp+1-argc;
        pc = p[pc]; 
      } break; 
      case TCALL: { 
        int argc = p[pc++];                // Number of new arguments
        int pop  = p[pc++];		   // Number of variables to discard
        for (int i=argc-1; i>=0; i--)	   // Discard variables
          s[sp-i-pop] = s[sp-i];
        sp = sp - pop; pc = p[pc]; 
      } break; 
      case RET: { 
        int res = s[sp]; 
        sp = sp-p[pc]; bp = s[--sp]; pc = s[--sp]; 
        s[sp] = res; 
      } break; 
      case PRINTI:
        System.out.print(s[sp] + " "); break; 
      case PRINTC:
        for(int i=0;i<=sp;i++){
          System.out.print(s[i]+" ");
        }
        System.out.print("\n");
        System.out.print(s[sp]+" "+(char)(s[sp])); break; 
      case PRINTF:{
        float ans=(float)change(s[sp]);
        System.out.print(ans +" "); break;
      }
        
      case LDARGS:
	for (int i=0; i<iargs.length; i++) // Push commandline arguments
	  s[++sp] = iargs[i];
	break;
      case STOP:
        return sp;
      default:                  
        throw new RuntimeException("Illegal instruction " + p[pc-1] 
                                   + " at address " + (pc-1));
      }
    }
  }

  // Print the stack machine instruction at p[pc]
  static String insname(int[] p, int pc) {
    switch (p[pc]) {
    case CSTI:   return "CSTI " + p[pc+1]; 
    case CSTF:   return "CSTF " + p[pc+1]; 
    case CSTS:   return "CSTS " + p[pc+1]+" "+p[pc+2]; 
    case BITNOT:  return "BITNOT";
    case BITLEFT: return "BITLEFT";
    case BITRIGHT:  return "BITRIGHT";
    case BITAND: return "BITAND";
    case BITOR:  return "BITOR";
    case BITXOR: return "BITXOR";
    case ADD:    return "ADD";
    case SUB:    return "SUB";
    case MUL:    return "MUL";
    case DIV:    return "DIV";
    case MOD:    return "MOD";
    case EQ:     return "EQ";
    case LT:     return "LT";
    case NOT:    return "NOT";
    case DUP:    return "DUP";
    case SWAP:   return "SWAP";
    case LDI:    return "LDI";
    case STI:    return "STI";
    case GETBP:  return "GETBP";
    case GETSP:  return "GETSP";
    case INCSP:  return "INCSP " + p[pc+1];
    case GOTO:   return "GOTO " + p[pc+1];
    case IFZERO: return "IFZERO " + p[pc+1];
    case IFNZRO: return "IFNZRO " + p[pc+1];
    case CALL:   return "CALL " + p[pc+1] + " " + p[pc+2];
    case TCALL:  return "TCALL " + p[pc+1] + " " + p[pc+2] + " " + p[pc+3];
    case RET:    return "RET " + p[pc+1];
    case PRINTI: return "PRINTI";
    case PRINTC: return "PRINTC";
    case LDARGS: return "LDARGS";
    case STOP:   return "STOP";
    case NEG:    return "NEG";
    case INVO:   return "INVO";
    case GCD:    return "GCD";
    default:     return "<unknown>";
    }
  }

  // Print current stack and current instruction

  static void printsppc(int[] s, int bp, int sp, int[] p, int pc) {
    System.out.print("[ ");
    for (int i=0; i<=sp; i++)
      System.out.print(s[i] + " ");
    System.out.print("]");
    System.out.println("{" + pc + ": " + insname(p, pc) + "}"); 
  }

  // Read instructions from a file

  public static int[] readfile(String filename) 
    throws FileNotFoundException, IOException
  {
    ArrayList<Integer> rawprogram = new ArrayList<Integer>();
    Reader inp = new FileReader(filename);
    StreamTokenizer tstream = new StreamTokenizer(inp);
    tstream.parseNumbers();
    tstream.nextToken();
    while (tstream.ttype == StreamTokenizer.TT_NUMBER) {
      rawprogram.add(new Integer((int)tstream.nval));
      tstream.nextToken();
    }
    inp.close();
    final int programsize = rawprogram.size();
    int[] program = new int[programsize];
    for (int i=0; i<programsize; i++)
      program[i] = ((Integer)(rawprogram.get(i))).intValue();
    return program;
  }
}

// Run the machine with tracing: print each instruction as it is executed

class Machinetrace {
  public static void main(String[] args)        
    throws FileNotFoundException, IOException {
    if (args.length == 0) 
      System.out.println("Usage: java Machinetrace <programfile> <arg1> ...\n");
    else
      Machine.execute(args, true);
  }
}
